### PR TITLE
Added back button confirmation for quitting the app

### PR DESCRIPTION
**Pull request by valentine-bvc**
Like mentioned in the thread i have implemented the feature in all 3 views and it works perfectly according to my testing on windows, android virtual device (android 14) and android smarthphone (android 7) on flutter version 3.3. Updated the readme in the commit too, and added troubleshooting for setting up the projects and dealing with some errors!

### CODE REVIEW

File: android/app/build.gradle [hunk 1]
```diff
@@ -8,7 +8,8 @@ if (localPropertiesFile.exists()) {
 
 def flutterRoot = localProperties.getProperty('flutter.sdk')
 if (flutterRoot == null) {
-    throw new GradleException("Flutter SDK not found. Define location with flutter.sdk in the local.properties file.")
+   throw new GradleException("Flutter SDK not found. Define location with flutter.sdk in the local.properties file.")
```
Comment (joneldominic) [inline android/app/build.gradle]: Are the following changes automatically generated or manually changed? If manually changed, please remove unnecessary changes.

File: lib/app/app.locator.dart
Comment (joneldominic) [inline lib/app/app.locator.dart:1]: It looks like the following changes are unnecessary. Please run `flutter format . -l 100` then add it as a separate commit.

Comment (joneldominic) [reply lib/app/app.locator.dart:1]: The above comment applies to all files.

File: README.md [hunk 2]
```diff
@@ -111,6 +111,18 @@ Light Mode            |  Dark Mode
...
+5. Troubleshooting:
+   - if getting errors when building try
+   ```sh
+   flutter pub upgrade
+   ```
+   - if getting errors when trying to build release apk with "flutter build apk --release" try
+   ```sh
+   flutter build apk --release --no-tree-shake-icons
+   ```
+   - if you're getting "missing whatsap icons" when trying to runn the app, it could be because the newest flutter version is incompatible, if so, you need to downgrade your flutter version to an older build and try again. Version 3.3 was tested and works as intended.
```
Comment (joneldominic) [inline README.md]: Apply the following:

- Update `whatsap` into `whatsapp`
- Update `runn` into `run`

File: lib/ui/views/settings/settings_view.dart
Comment (joneldominic) [inline lib/ui/views/settings/settings_view.dart:1]: `MainView` already contains Scaffold ([Line 36](https://github.com/joneldominic/personal-finance-management-app/pull/83/files#diff-fec8272e85fa1a6c6d018fc56b02da963c021b544d408c45aa19eb3a20dfbf2eR36)), can you apply the `WillPopScope` there? This way we can reduce code redundancy. Also, it would be better to put our logic in the `ViewModel` (ex. `MainViewModel`). Lastly, you may wanna reuse `_snackbarService.showCustomSnackBar` to show the snackbar.

**Review by joneldominic — CHANGES_REQUESTED** [review]

File: android/app/build.gradle (hunk 1 above)
Comment (valentine-bvc) [reply android/app/build.gradle]: had a little error with this but found out afterwards that it's safe to ignore the error as it's vs code related or something and we shouldn't mess with the way it throws exception, will rectify that

File: lib/ui/views/settings/settings_view.dart
Comment (valentine-bvc) [reply lib/ui/views/settings/settings_view.dart:1]: I did wonder if there was a way to do it more efficiently rather that have duplicated code for every view but my limited flutter knowledge gave me a hard time haha

**Comment by valentine-bvc** [comment]
I will address all the changes requested the best I can and let you have another look soon. Hopefully today or tomorrow!

**Comment by valentine-bvc** [comment]
Update: I addressed most requests, i have reformatted the files, fixed the typos, i migrated the WillPopScope to MainView and put the logic in MainViewModel which is much nicer now. But I am running in on a bug with your latest request. I am trying to use _snackbarService.showCustomSnackBar instead to display the snackbar in my handleWillPop and the problem is the snackbar doesn't appear at all until i change the screen (example back button press once on home screen then go to settings, that's when the snackbar appears instead of appearing on the home screen) and it only happens when I use your custom snackbar, maybe i need to look more into it but I thought I'd ask, maybe you have a suggestion?

**Comment by valentine-bvc** [comment]
hey! sorry for no more updates, i haven't had a chance to play more with the _snackbarService.showCustomSnackBar since i've been pretty busy, hopefully i get more time soon and get figure it out! Hope all is good with you!

**Comment by joneldominic** [comment]
> Update: I addressed most requests, i have reformatted the files, fixed the typos, i migrated the WillPopScope to MainView and put the logic in MainViewModel which is much nicer now. But I am running in on a bug with your latest request. I am trying to use _snackbarService.showCustomSnackBar instead to display the snackbar in my handleWillPop and the problem is the snackbar doesn't appear at all until i change the screen (example back button press once on home screen then go to settings, that's when the snackbar appears instead of appearing on the home screen) and it only happens when I use your custom snackbar, maybe i need to look more into it but I thought I'd ask, maybe you have a suggestion?

> hey! sorry for no more updates, i haven't had a chance to play more with the _snackbarService.showCustomSnackBar since i've been pretty busy, hopefully i get more time soon and get figure it out! Hope all is good with you!

Hello, sorry. I've got diverted into other priorities lately. 

No worries. I've also reproduced the issue you've mentioned and I still have no idea how to make it work. I'll try to explore more regarding the issue.

File: lib/ui/views/main/main_view.dart [hunk 3]
```diff
@@ -18,40 +18,45 @@ import 'main_viewmodel.dart';
...
-  const MainView({Key? key}) : super(key: key);
+  // ignore: prefer_const_constructors_in_immutables
+  MainView({Key? key}) : super(key: key);
 
   @override
   Widget build(BuildContext context) {
     final customTheme = Theme.of(context).extension<CustomTheme>()!;
 
     const int defaultViewIndex = 1;
+    MainViewModel mainViewModel = MainViewModel();
```
Comment (joneldominic) [inline lib/ui/views/main/main_view.dart]: It's not necessary to have another instance of `MainViewModel`. You can just reuse the `model` found in Line 36 below. `builder: (context, model, child) {...`

File: lib/ui/views/main/main_viewmodel.dart [hunk 4]
```diff
@@ -52,4 +53,33 @@ class MainViewModel extends MultipleStreamViewModel {
     _logger.i('argument: NONE');
     _navigationService.navigateToTransactionDetailView();
   }
+
```
Comment (joneldominic) [inline lib/ui/views/main/main_viewmodel.dart:56]: I think, we could just time block on making the `_snackbarService.showCustomSnackBar` work. If we really can't find a solution we can just settle with this approach. But I would suggest to make the UI to look similar with other snackbars. (For example when adding a transaction with `0` amount)

**Review by joneldominic — CHANGES_REQUESTED** [review]
Please address new comments.

File: lib/ui/views/settings/settings_view.dart
Comment (joneldominic) [reply lib/ui/views/settings/settings_view.dart:1]: No worries. It's part of the learning process 👌

File: lib/ui/views/main/main_view.dart (hunk 3 above)
Comment (valentine-bvc) [reply lib/ui/views/main/main_view.dart]: ops i will look into this and fix it! sorry

File: lib/ui/views/main/main_viewmodel.dart (hunk 4 above)
Comment (valentine-bvc) [reply lib/ui/views/main/main_viewmodel.dart:56]: i will look into doing this, and yeah i've seen the other snackbars, will test around and see what can be done! cheers!

**Comment by valentine-bvc** [comment]
Hello! Okay so i managed to spend a couple of hours on this and the best i was able to do is make it match your snackbar's colors, i wanted to copy your snackbar including the round corners and smaller dimensions and what not but i don't think i can do it. The closest i got is by changing the behavior of the snackbar to a floating type, which makes it pretty similar to yours dimensions wise,but then the problem is that the snackbar will not be at the bottom anymore, as it will stay on top of any floating actions buttons like the ones you have such as add transaction and what not, and this cannot be forced to be at the bottom according to the documentation I've seen. I've tried padding the normal snackbar also, but doesn't help. The only real solution would be to make another custom snackbar which i suppose would be a bit of a mess as i would do what you already did and will probably run into the delay bug again. I can keep on looking maybe i'm missing something but this is what i came up with so far. I've matched the colors and addressed the other problem!

**Comment by valentine-bvc** [comment]
Little update: So yeah the stock snackbar is very limited, I cannot change the margins of width of it, i can make the corners a little rounder but overall is not gonna be smaller like yours is, as the snackbar won't allow changing these unless I make the snackbar a floating type which will conflict with your floating action buttons making the snackbar pop in the middle of the screen instead of the bottom, so the best I can do with a stock snackbar is match the colors.
I came across a third party snackbar https://pub.dev/packages/flushbar but I don't know if you'd be interested in this as you already have your own custom. I can push the changes with the modelview correction and color matching but like I said, in terms of stock snackbar with chaning it's size and making it a little smaller just like yours, I think I've exhausted all my options.

**Comment by joneldominic** [comment]
> Little update: So yeah the stock snackbar is very limited, I cannot change the margins of width of it, i can make the corners a little rounder but overall is not gonna be smaller like yours is, as the snackbar won't allow changing these unless I make the snackbar a floating type which will conflict with your floating action buttons making the snackbar pop in the middle of the screen instead of the bottom, so the best I can do with a stock snackbar is match the colors. I came across a third party snackbar https://pub.dev/packages/flushbar but I don't know if you'd be interested in this as you already have your own custom. I can push the changes with the modelview correction and color matching but like I said, in terms of stock snackbar with chaning it's size and making it a little smaller just like yours, I think I've exhausted all my options.

Have you tried using flushbar? Does it work fine?

**Comment by valentine-bvc** [comment]
> Have you tried using flushbar? Does it work fine?

I haven't as i wanted your input first. Will test it out and let you know if it's a better option.

**Comment by valentine-bvc** [comment]
Update: okay so forget about original flushbar as it is deprecated and apparently may not be supporting flutter 3.0 but i found a new iteration called another_flushbar ( https://pub.dev/packages/another_flushbar ) and i got it to work as well as replicate your custom snackbar pretty well. It also has a similar positioning it does appear when it's supposed to instead of being delayed. I did keep the text left sided instead of centered after all because it is more consistent with your custom snackbar. But you tell me what you think. I haven't pushed anything yet but here is a quick pic: https://imgur.com/a/iTxZdN0

**Comment by joneldominic** [comment]
> Update: okay so forget about original flushbar as it is deprecated and apparently may not be supporting flutter 3.0 but i found a new iteration called another_flushbar ( https://pub.dev/packages/another_flushbar ) and i got it to work as well as replicate your custom snackbar pretty well. It also has a similar positioning it does appear when it's supposed to instead of being delayed. I did keep the text left sided instead of centered after all because it is more consistent with your custom snackbar. But you tell me what you think. I haven't pushed anything yet but here is a quick pic: https://imgur.com/a/iTxZdN0

Hello @valentine-bvc . I've just checked the screenshot you've attached. I'm good with it. Please push your changes so I can review it again. 

Note: Sorry for the very delayed response.

**Comment by joneldominic** [comment]
Hello @valentine-bvc , I've manage to make SnackbarService working without any delay. You might wanna check my changes on branch [feature/66-add-back-button-confirmation-working](https://github.com/joneldominic/personal-finance-management-app/compare/feature/66-add-back-button-confirmation-working?expand=1). Please apply those changes in your PR, together with all your the necessary changes (ex. Troubleshooting guide, etc.).

Edit: We just miss adding the `instantInit`  [flag](https://github.com/joneldominic/personal-finance-management-app/compare/feature/66-add-back-button-confirmation-working?expand=1#diff-dc4200ad7df71884a41eefc7c9f88d9678ae34c15c7f5f66e74d50ac88830ddcR14) when setting up the snackbar

**Comment by valentine-bvc** [comment]
Gotcha, will test this out asap!

**Comment by joneldominic** [comment]
> Gotcha, will test this out asap!

Another thing @valentine-bvc... It's a good practice to match Branch Name or PR with the actual issue number (https://github.com/joneldominic/personal-finance-management-app/issues/66) so that it can easily tracked or probably github will link it automatically. 

So I suggest creating a new branch with the following name... `feature/66-add-back-button-confirmation` then create a new PR.

**Comment by valentine-bvc** [comment]
Alright in this case i will also just build on top of your last commit instead of just merging just to avoid the mess because there already are too many useless commits such as added my own snackbar, format, etc . I will just test out your fix, make it all in one or two nice commits and make a pull request. I already spent a lot of time on this so i should have a pretty easy time figuring things out.

**Comment by joneldominic** [comment]
> Alright in this case i will also just build on top of your last commit instead of just merging just to avoid the mess because there already are too many useless commits such as added my own snackbar, format, etc . I will just test out your fix, make it all in one or two nice commits and make a pull request. I already spent a lot of time on this so i should have a pretty easy time figuring things out.

Yep, branching out from develop branch much would be better.
